server/app/assets/javascripts/file_upload.ts

```diff
@@ -53,6 +56,62 @@ export function init() {
       blockForm.submit()
     }
   })
+
+  const uploadedFilesAttribute = blockForm
+    .querySelector(`[${UPLOADED_FILE_ATTR}]`)
+    ?.getAttribute(UPLOADED_FILE_ATTR)
+
+  if (uploadedFilesAttribute) {
+    const uploadedFilesArray = JSON.parse(uploadedFilesAttribute) as string[]
+    blockForm.addEventListener('formdata', (event) => {
+      const formData = event.formData
+      const file = formData.get('file') as File
+
+      const newName = getUniqueName(file.name, uploadedFilesArray)
+      if (file.name != newName) {
+        // Rename uploaded file, if a file with the same name has already been uploaded.
+        formData.delete('file')
+        formData.append('file', file, newName)
+      }
+    })
+  }
+}
+
+// Exporting so we can test this.
+export function getUniqueName(name: string, existingNames: string[]) {
+  while (existingNames.includes(name)) {
+    const fileNameWithDigitSuffixMatch = name.match(
+      FILE_NAME_DIGIT_SUFFIX_REGEX,
+    )
+    let numberToAppend = 2
+    if (fileNameWithDigitSuffixMatch) {
+      // File name already has a digit suffix. Increment that instead.
+      if (fileNameWithDigitSuffixMatch[2]) {
+        numberToAppend =
+          parseInt(fileNameWithDigitSuffixMatch[2].substring(1)) + 1
+      }
+      name =
+        fileNameWithDigitSuffixMatch[1] +
+          '-' +
+          numberToAppend +
+          fileNameWithDigitSuffixMatch[3] ?? ''
```

Comment:
question (non-blocking): what might `fileNameWithDigitSuffixMatch[3]` return?

Reply:
It returns the file extension (.png), or null if there isn't one. (I also added documentation like you suggested)